### PR TITLE
Function: Introduce `Function` as a separate sub-class to Subroutine

### PR DESCRIPTION
**Pull request by mlange05**
_Note: This sits on top of PR #550 and should only be reviewed once this is merged. Filing the PR now for testing only._

This PR refactors the basic procedure handling to separate `Function` as a distinct sub-class of `Subroutine` that now exclusively handle the properties `result_name` and `return_type`. The `Subroutine` still provides the `is_function` attribute, but merely as a class attribute that is always false.

The PR also separates function-specific tests concerning return types and prefix/suffix checks from the subroutine tests and adds dedicated prefix/suffix tests for `Subroutine`. To facilitate this, the code paths for `Subroutine` / `Function` in frontend and backend are now replicated, as the there are very subtle differences in the constructor and return type handling, despite the large amount of other commonalities.

Importantly, this now also removes the (seemingly redundant) explicit addition of a `VariablesDeclaration` node for function return types without explicit return type declarations. A new test has been added to cover the correct handling of the different types of function return type definitions (I'm sure I've missed some corner cases here 😉 ).

### CODE REVIEW

**Review by reuterbal — CHANGES_REQUESTED** [review]
MANY thanks for this, a very important stepping stone towards #196! I've left a few comments but all of them rather minor or even incorrect, so please have look if any of these make sense.

Coincidentally, might this also provide a fix for #513 already?

File: loki/frontend/fparser.py [hunk 1]
```diff
@@ -1890,29 +1869,153 @@ def visit_Subroutine_Subprogram(self, o, **kwargs):
...
+            body = sanitize_ir(body, FP, pp_registry=sanitize_registry[FP], pp_info=self.pp_info)
+
+        # Workaround for lost StatementFunctions:
+        # Since FParser has no means to identify StmtFuncs, the last set of them
+        # can get lumped in with the body, and we simply need to shift them over.
+        stmt_funcs = tuple(n for n in body.body if isinstance(n, ir.StatementFunction))
+        if stmt_funcs:
+            idx = body.body.index(stmt_funcs[-1]) + 1
+            spec._update(body=spec.body + body.body[:idx])
+            body._update(body=body.body[idx:])
```
[the remark this answers is not in the document]
Comment (mlange05) [reply loki/frontend/fparser.py:1960]: Yes, ultimately I'd like to create an overarching abstract `Procedure` type that inherits from `ProgramUnit` for scope/variable handling, but adds effectively argument handling and all things shared in the current step. But this gets us quickly into "procedures as scope nodes" territory, so for now I want to cleanly separate them first.

File: loki/backend/fgen.py [hunk 2]
```diff
@@ -203,19 +203,63 @@ def visit_Module(self, o, **kwargs):
...
         """
         ftype = 'FUNCTION' if o.is_function else 'SUBROUTINE'
         prefix = self.join_items(o.prefix, sep=' ')
         if o.prefix:
             prefix += ' '
         arguments = self.join_items(o.argnames)
+        if isinstance(o.bind, str):
+            bind_c = f' BIND(c, name="{o.bind}")'
+        elif isinstance(o.bind, StringLiteral):
+            bind_c = f' BIND(c, name={o.bind})'
```
[the remark this answers is not in the document]
Comment (mlange05) [reply loki/backend/fgen.py:221]: `StringLiteral` is the native representation created by the frontends, but many of the "bind" interface routines are created in the transpilation pipelines, and likely not all use `StringLiteral`s. The case itself was covered, but it was not invoking `fgen`, so I've now added this to the test.

**Comment by mlange05** [comment]
Ok, I think I either addressed or answered all comments. Over to you, @reuterbal . 😉

File: loki/transformations/build_system/tests/test_dependency.py [hunk 3]
```diff
@@ -683,7 +683,7 @@ def test_dependency_transformation_inline_call(frontend):
     assert kernel['kernel_test_mod'] == kernel.modules[0]
 
     # Check that the return name hasn't changed
-    assert 'kernel' in kernel['kernel_test'].variables
+    assert 'kernel' in kernel['kernel_test'].symbol_attrs
```
[the remark this answers is not in the document]
Comment (reuterbal) [reply loki/transformations/build_system/tests/test_dependency.py:686]: OK, understood. I was wondering if longer term the behaviour of the `variables` property (and thus implicitly all derived properties like `variable_map`) should inject the return variable into the list for functions with this implicit declaration. It requires to then also filter this again in the setter method - so definitely requires some acrobatics but sounds not entirely impossible to me.

Definitely not necessary for this PR, just curious what your thoughts are on this.
